Expand site isolation testing to cover timing of _webView:didCommitLoadWithRequest:inFrame: callback

achristensen07 opened the pull request:
#### 591ba3643d0068fd9cfbbef10586f7048e8771cf
<pre>
Expand site isolation testing to cover timing of _webView:didCommitLoadWithRequest:inFrame: callback
<a href="https://bugs.webkit.org/show_bug.cgi?id=249047">https://bugs.webkit.org/show_bug.cgi?id=249047</a>

Reviewed by Chris Dumez.

Also write 2 lines of code instead of using std::exchange with Ref, which asan doesn&apos;t seem to like right now.

* Source/WebKit/UIProcess/WebFrameProxy.cpp:
(WebKit::WebFrameProxy::commitProvisionalFrame):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolation.mm:
(TestWebKitAPI::TEST):
* Tools/TestWebKitAPI/cocoa/TestNavigationDelegate.h:
* Tools/TestWebKitAPI/cocoa/TestNavigationDelegate.mm:
(-[TestNavigationDelegate _webView:didCommitLoadWithRequest:inFrame:]):

Canonical link: <a href="https://commits.webkit.org/257742@main">https://commits.webkit.org/257742@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1094a48e1cb7d33fc6ec44ddb6a0012ec11436d7

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/99664 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/8841 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/32744 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/109026 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/169257 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/103668 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/9427 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/86130 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/92125 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/106934 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/105436 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/7105 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/90618 "Passed tests") | [💥 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/34064 "An unexpected error occured. Recent messages:Cleaned up git repository; Skipping applying patch since patch_id isn't provided; Checked out pull request; Updated gtk dependencies; extracting built product; Pull request contains relevant changes; Skipped layout-tests; layout-tests (exception)") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/88920 "Passed tests") | [❌ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/21977 "Found 3 new test failures: media/modern-media-controls/media-controls/media-controls-display-above-captions.html, media/modern-media-controls/placard-support/placard-support-airplay-fullscreen.html, media/modern-media-controls/status-support/status-support-error.html (failure)") | [⏳ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/API-Tests-GTK-EWS "Waiting to run tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/2664 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/23493 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/2609 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/45876 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/8746 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/42967 "Passed tests") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/2724 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/4471 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebKit/UIProcess/WebFrameProxy.cpp:
```diff
@@ -400,7 +400,8 @@ void WebFrameProxy::commitProvisionalFrame(FrameIdentifier frameID, FrameInfoDat
     send(Messages::WebFrame::DidCommitLoadInAnotherProcess());
     m_process->removeMessageReceiver(Messages::WebFrameProxy::messageReceiverName(), m_frameID.object());
-    m_parentFrameProcess = std::exchange(m_process, m_provisionalFrame->process());
+    m_parentFrameProcess = m_process.ptr();
```
Either the naming is wrong or the logic is wrong. Currently, this is not the parent frame process but the process before the last swap.

Sure, on the first swap, this happens to be the parent frame process. However, on later swap, this is no longer true.

achristensen07 replied:
Naming is probably not ideal, but this is the process which hosts the parent frame if it is different than this frame.  It is all likely to be considerably reworked in the coming weeks.

cdumez replied:
But this is not true, it is the process that was used by the frame before the *last* process swap. It is not necessarily the parent frame process. I think you should initialize m_parentFrameProcess differently.

Presumably, you could just initialize m_parentFrameProcess on WebFrameProcess construction to be the same as m_process for example. And then no worry about updating it later?

cdumez replied:
Or not overwrite `m_parentFrameProcess` if it is already set. If you don't want to eagerly initialize m_parentFrameProcess in the constructor for some reason.

achristensen07 replied:
I'm going to remove this change from this PR and defer this conversation.  This PR is about the test.

achristensen07 committed:
Expand site isolation testing to cover timing of _webView:didCommitLoadWithRequest:inFrame: callback
https://bugs.webkit.org/show_bug.cgi?id=249047

Reviewed by Chris Dumez.

Also write 2 lines of code instead of using std::exchange with Ref, which asan doesn't seem to like right now.

* Source/WebKit/UIProcess/WebFrameProxy.cpp:
(WebKit::WebFrameProxy::commitProvisionalFrame):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolation.mm:
(TestWebKitAPI::TEST):
* Tools/TestWebKitAPI/cocoa/TestNavigationDelegate.h:
* Tools/TestWebKitAPI/cocoa/TestNavigationDelegate.mm:
(-[TestNavigationDelegate _webView:didCommitLoadWithRequest:inFrame:]):

Canonical link: https://commits.webkit.org/257742@main
